### PR TITLE
Remove renter's merkle roots from ram

### PR DESCRIPTION
**Pull request by ChrisSchinnerl**
Instead of having all the merkle roots of every contract in memory, the renter should load them from disk if needed.

### CODE REVIEW

**Comment by ChrisSchinnerl** [comment]
This PR is dependent on [PR#19](https://github.com/NebulousLabs/merkletree/pull/19)

File: modules/renter/proto/contract.go [hunk 1]
```diff
@@ -345,8 +337,9 @@ func (cs *ContractSet) managedInsertContract(h contractHeader, roots []crypto.Ha
 		return modules.RenterContract{}, err
 	}
 	// write roots
-	for i, root := range roots {
-		if _, err := f.WriteAt(root[:], contractHeaderSize+crypto.HashSize*int64(i)); err != nil {
+	merkleRoots := newMerkleRoots(f)
+	for _, root := range roots {
+		if err := merkleRoots.push(root); err != nil {
```
Comment (DavidVorick) [inline modules/renter/proto/contract.go]: need to add a TODO somewhere in this package that currently the startup routine involves reading all of the merkle roots into memory, then hashing all of the merkle roots in the the `merkleRoots` object. That's a slow process that we should fix at some point, but it's probably not necessary until after we fix up the file structures.

File: modules/renter/proto/merkleroots.go [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+		// cached subTree in cachedSubTrees.
+		uncachedRoots []crypto.Hash
+
+		// file is the file of the safe contract that contains the root.
+		file *os.File
+		// numMerkleRoots is the number of merkle roots in file.
+		numMerkleRoots int
+	}
+
+	cachedSubTree struct {
```
Comment (DavidVorick) [inline modules/renter/proto/merkleroots.go]: need docstring

File: modules/renter/proto/merkleroots.go [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+// loadExistingMerkleRoots reads creates a merkleRoots object from existing
+// merkle roots.
+func loadExistingMerkleRoots(file *os.File) (mr *merkleRoots, err error) {
+	mr = &merkleRoots{
+		file: file,
+	}
+	// Get the number of roots stored in the file.
+	mr.numMerkleRoots, err = mr.lenFromFile()
+	if err != nil {
+		return
```
Comment (DavidVorick) [inline modules/renter/proto/merkleroots.go]: convention would be to return `nil, err` anytime that there's an error, same follows for the rest of the function. For a function this long, it's we tend to strongly prefer unnamed return values, so that we have to explicitly state what is being returned.

File: modules/renter/proto/merkleroots.go [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+	}
+	// Seek to the first root's offset.
+	if _, err = file.Seek(rootIndexToOffset(0), io.SeekStart); err != nil {
+		return
+	}
+	// Read the roots from the file without reading all of them at once.
+	for i := 0; i < mr.numMerkleRoots; i++ {
```
Comment (DavidVorick) [inline modules/renter/proto/merkleroots.go]: I would add a TODO at the top of the file, reading the roots one at a time is pretty inefficient. The OS will handle a lot of that via disk caching, but I'm guessing we would see some i/o and performance improvements if we read say, 100,000 roots at a time instead of one at a time.

No need to do that in this PR, I'd just write a comment highlighting that it's something to do.

Actually, this is the type of optimization that would be really great to throw at a newer developer like Matt or one of the guys in the #core-dev channel.

Comment (DavidVorick) [reply modules/renter/proto/merkleroots.go]: @tbenz9 maybe we should make an issue when this is merged and get someone in the open source community to look at it.

File: modules/renter/proto/merkleroots.go [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+		var root crypto.Hash
+		if _, err = io.ReadFull(file, root[:]); err == io.EOF {
+			break
+		} else if err != nil {
+			return
+		}
+
+		// Append the root to the unachedRoots
```
Comment (DavidVorick) [inline modules/renter/proto/merkleroots.go]: unachedRoots. I guess they used to ache, but now they have been unached :P

File: modules/renter/proto/merkleroots.go [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+		mr.uncachedRoots = append(mr.uncachedRoots, root)
+
+		// If the uncachedRoots grew too large we add them to the cache.
+		if len(mr.uncachedRoots) == (1 << merkleRootsCacheHeight) {
```
Comment (DavidVorick) [inline modules/renter/proto/merkleroots.go]: This code should be abstracted, there should be a function for adding a new root to the data structure. I know we have the insert() function, but that one writes to disk as well as adding to the in-memory structure. Here we only care about adding to the in-memory structure, but that should still be its own abstraction that gets called anywhere the in-memory structure is extended (including insert())

File: modules/renter/proto/merkleroots.go [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+		}
+	}
+	return mr, nil
+}
+
+// newCachedSubTree creates a cachedSubTree from exactly
+// 2^merkleRootsCacheHeight roots.
+func newCachedSubTree(roots []crypto.Hash) *cachedSubTree {
+	// Sanity check the input length.
+	if len(roots) != (1 << merkleRootsCacheHeight) {
```
Comment (DavidVorick) [inline modules/renter/proto/merkleroots.go]: So, height of 0 means it's the hash of a leaf, height of 1 means it's the root of 2 leaves, height of 2 means root of 4 leaves, etc?

I don't think that's clearly documented in the merkle tree cached functions, we should make sure the docstrings for all of the cached roots properly explain what counts as height 0 and 1 etc.

File: modules/renter/proto/merkleroots.go [hunk 8]
```diff
@@ -0,0 +1,343 @@
...
+// Don't use this on a file that contains roots.
+func newMerkleRoots(file *os.File) *merkleRoots {
+	return &merkleRoots{
+		file: file,
+	}
+}
+
+// rootIndexToOffset calculates the offset of the merkle root at index i from
+// the beginning of the contract file.
+func rootIndexToOffset(i int) int64 {
```
Comment (DavidVorick) [inline modules/renter/proto/merkleroots.go]: this function name took me a few reads to understand what it was saying. I understand it now, but maybe there's a different name which isn't as confusing on the first read.

File: modules/renter/proto/merkleroots.go [hunk 9]
```diff
@@ -0,0 +1,343 @@
...
+		// persisted and are computed after startup.
+		cachedSubTrees []*cachedSubTree
+		// uncachedRoots contains the sector roots that are not part of a
+		// cached subTree. The uncachedRoots slice should never get longer than
+		// 2^merkleRootsCacheHeight since that would simply result in a new
+		// cached subTree in cachedSubTrees.
+		uncachedRoots []crypto.Hash
+
+		// file is the file of the safe contract that contains the root.
+		file *os.File
```
Comment (lukechampine) [inline modules/renter/proto/merkleroots.go]: I don't like that both `merkleRoots` and `SafeContract` have a handle to the same file. If you're just looking at the `SafeContract` code, it's not obvious that calling a method on `sc.merkleRoots` modifies sc.f`. I'm not sure what the best way to resolve this is. Maybe it's a mistake to have a separate `merkleRoots` type, and it should just be merged with `SafeContract`.

File: modules/renter/proto/merkleroots.go (hunk 4 above)
Comment (lukechampine) [reply modules/renter/proto/merkleroots.go]: optimizing this is really simple thanks to `bufio`:
```go
r := bufio.NewReader(file)
```
You can get a little faster by using a custom (larger) buffer size, but the default buffer size will probably get you to 90% peak performance.

File: modules/renter/proto/merkleroots.go [hunk 10]
```diff
@@ -0,0 +1,343 @@
...
+	return i / (1 << merkleRootsCacheHeight), true
+}
+
+// lenFromFile returns the number of merkle roots by computing it from the
+// filesize.
+func (mr *merkleRoots) lenFromFile() (int, error) {
+	offset, err := mr.file.Seek(0, io.SeekEnd)
+	if err != nil {
+		return 0, err
+	}
```
Comment (lukechampine) [inline modules/renter/proto/merkleroots.go]: more idiomatic to get the size via `Stat`:
```go
stat, err := mr.file.Stat()
if err != nil {
    return 0, err
}
size := stat.Size()
```

File: modules/renter/proto/merkleroots.go [hunk 11]
```diff
@@ -0,0 +1,343 @@
...
+	// If we haven't written a single root yet we just return 0.
+	rootStart := rootIndexToOffset(0)
+	if offset < rootStart {
+		return 0, nil
+	}
+
+	// Sanity check contract file length.
+	if (offset-rootStart)%crypto.HashSize != 0 {
+		build.Critical("contract file has unexpected length and might be corrupted.")
```
Comment (lukechampine) [inline modules/renter/proto/merkleroots.go]: this is not deserving of a `build.Critical`. The contract could be truncated to the nearest hash, or it could simply return an error and treat the contract as unusable -- but it's not an unrecoverable error.

File: modules/renter/proto/merkleroots.go [hunk 12]
```diff
@@ -0,0 +1,343 @@
...
+
+// lenFromFile returns the number of merkle roots by computing it from the
+// filesize.
+func (mr *merkleRoots) lenFromFile() (int, error) {
+	offset, err := mr.file.Seek(0, io.SeekEnd)
+	if err != nil {
+		return 0, err
+	}
+	// If we haven't written a single root yet we just return 0.
+	rootStart := rootIndexToOffset(0)
```
Comment (lukechampine) [inline modules/renter/proto/merkleroots.go]: isn't this just `contractHeaderSize`? Not sure if one is more clear than the other.

Comment (lukechampine) [reply modules/renter/proto/merkleroots.go]: also, is this branch actually necessary? Won't the remaining logic return 0 just fine?

File: modules/renter/proto/merkleroots.go [hunk 13]
```diff
@@ -0,0 +1,343 @@
...
+// number as lenFromFile.
+func (mr *merkleRoots) len() int {
+	return mr.numMerkleRoots
+}
+
+// push appends a merkle root to the end of the contract. If the number of
+// uncached merkle roots grows too big we cache them in a new subTree.
+func (mr *merkleRoots) push(root crypto.Hash) error {
+	// Sanity check the number of uncached roots before adding a new one.
+	if len(mr.uncachedRoots) == (1 << merkleRootsCacheHeight) {
```
Comment (lukechampine) [inline modules/renter/proto/merkleroots.go]: `(1 << merkleRootsCacheHeight)` is a common enough expression that we may want to define a `merkleRootsCacheSize` or `merkleRootsPerCache` const.

File: modules/renter/proto/merkleroots.go [hunk 14]
```diff
@@ -0,0 +1,343 @@
...
+
+// merkleRootsFrom index readds all the merkle roots in range [from;to)
+func (mr *merkleRoots) merkleRootsFromIndex(from, to int) ([]crypto.Hash, error) {
+	merkleRoots := make([]crypto.Hash, 0, to-from)
+	if _, err := mr.file.Seek(rootIndexToOffset(from), io.SeekStart); err != nil {
+		return merkleRoots, err
+	}
+	for i := from; to-i > 0; i++ {
+		var root crypto.Hash
+		if _, err := io.ReadFull(mr.file, root[:]); err == io.EOF {
```
Comment (lukechampine) [inline modules/renter/proto/merkleroots.go]: ditto here on `bufio` optimization

File: modules/renter/proto/merkleroots_test.go [hunk 15]
```diff
@@ -0,0 +1,324 @@
...
+	roots, err = merkleRoots.merkleRoots()
+	if err != nil {
+		t.Fatal("failed to get roots from disk", err)
+	}
+	if roots[0] != newHash {
+		t.Fatal("root wasn't updated correctly on disk")
+	}
+	if merkleRoots.cachedSubTrees[0].sum != newCachedSubTree(roots[:1<<merkleRootsCacheHeight]).sum {
+		t.Fatal("cachedSubtree doesn't have expected sum")
+	}
```
Comment (DavidVorick) [inline modules/renter/proto/merkleroots_test.go]: should extend this test to include restarting / reloading to make sure that it's still consistent on disk after use.

File: modules/renter/proto/merkleroots_test.go [hunk 16]
```diff
@@ -0,0 +1,324 @@
...
+	// There should be 0 cached roots.
+	if len(merkleRoots.cachedSubTrees) != 0 {
+		t.Fatal("expected 0 cached roots but was", len(merkleRoots.cachedSubTrees))
+	}
+}
+
+// TestDeleteLastRoot tests the deleteLastRoot method by creating many roots
+// and deleting random indices until there are no more roots left.
+func TestDelete(t *testing.T) {
```
Comment (DavidVorick) [inline modules/renter/proto/merkleroots_test.go]: Docstring mismatch

File: modules/renter/proto/merkleroots_test.go [hunk 17]
```diff
@@ -0,0 +1,324 @@
...
+			subTreeLen := int(1 << merkleRootsCacheHeight)
+			from := cachedIndex * (1 << merkleRootsCacheHeight)
+			roots, err := merkleRoots.merkleRootsFromIndex(from, from+subTreeLen)
+			if err != nil {
+				t.Fatal("failed to read roots of subTree", err)
+			}
+			if merkleRoots.cachedSubTrees[cachedIndex].sum != newCachedSubTree(roots).sum {
+				t.Fatal("new cached root sum doesn't match expected sum")
+			}
+		}
```
Comment (DavidVorick) [inline modules/renter/proto/merkleroots_test.go]: performance might take too much of a hit if you do it every iteration, but maybe 1 in 100 iterations or something (randomly) you re-load the file and check for consistency.

File: modules/renter/proto/merkleroots_test.go [hunk 18]
```diff
@@ -0,0 +1,324 @@
...
+
+		// Insert
+		var hash crypto.Hash
+		copy(hash[:], fastrand.Bytes(len(hash)))
+		index := fastrand.Intn(merkleRoots.numMerkleRoots + 1)
+		if err := merkleRoots.insert(index, hash); err != nil {
+			t.Fatalf("failed to insert %v at %v: %v", hash, index, err)
+		}
+	}
```
Comment (DavidVorick) [inline modules/renter/proto/merkleroots_test.go]: occasional re-inits here as well I think

**Comment by DavidVorick** [comment]
Went through all the testing, mostly looks good. Want to make sure everything is being persisted correctly.

Are there any tests that would make sense to add to the `siatest` package? I want to make sure we have coverage for this, though I'm thinking right now that our existing `siatest` tests are probably good enough. The only other thing I can think that we'd need to check is the ACID guarantees, but we are working on that in a different set of pull requests.

File: modules/renter/proto/merkleroots_test.go [hunk 19]
```diff
@@ -11,6 +12,51 @@ import (
...
+	for i := 0; i < len(roots1); i++ {
+		if roots1[i] != roots2[i] {
+			return errors.New("on-disk roots don't match")
+		}
+	}
+	for i := 0; i < len(m1.uncachedRoots); i++ {
+		if m1.uncachedRoots[i] != m2.uncachedRoots[i] {
+			return errors.New("uncached roots don't match")
+		}
+	}
```
Comment (lukechampine) [inline modules/renter/proto/merkleroots_test.go]: `reflect.DeepEqual` makes these a bit cleaner

File: modules/renter/proto/merkleroots.go [hunk 20]
```diff
@@ -0,0 +1,359 @@
+// TODO currently the cached trees are not persisted and we build them at
+// startup. For petabytes of data this might take a long time.
```
Comment (DavidVorick) [inline modules/renter/proto/merkleroots.go]: File should take the form:

```go
// package asdf is an example package that doesn't exist.
pacakge asdf

// TODO: write the asdf package somewhere
```

File: modules/renter/proto/merkleroots.go (hunk 9 above)
Comment (DavidVorick) [reply modules/renter/proto/merkleroots.go]: going to agree with this, shouldn't have 2 objects with atomic assumptions pointing to the same file on disk, they might break eachother.

We should have this all happen under the SafeContract, using the wal of the SafeContract.

File: modules/renter/proto/merkleroots.go [hunk 21]
```diff
@@ -0,0 +1,359 @@
...
+		file: file,
+	}
+	// Get the number of roots stored in the file.
+	var err error
+	mr.numMerkleRoots, err = mr.lenFromFile()
+	if err != nil {
+		return nil, err
+	}
+	// Seek to the first root's offset.
+	if _, err = file.Seek(fileOffsetFromRootIndex(0), io.SeekStart); err != nil {
```
Comment (DavidVorick) [inline modules/renter/proto/merkleroots.go]: I'm thinking we want this function to be a method on the contract, instead of a global constant. I suspect that if we ever change the offset in the future, having it a method of the safecontract will clean things up when we transition from one thing to the next.

File: modules/renter/proto/merkleroots.go [hunk 22]
```diff
@@ -0,0 +1,359 @@
...
+		return nil, err
+	}
+	// Read the roots from the file without reading all of them at once.
+	r := bufio.NewReader(file)
+	for i := 0; i < mr.numMerkleRoots; i++ {
```
Comment (DavidVorick) [inline modules/renter/proto/merkleroots.go]: I'm thinking we should read more than 1 root at a time. I'm not sure what type of cacheing the OS does, but if we read 1000 or 10,000 roots at a time we'll probably see big speedups on systems without good OS level caching.

File: modules/renter/proto/merkleroots.go (hunk 21 above)
Comment (DavidVorick) [reply modules/renter/proto/merkleroots.go]: maybe it's not such a big deal

File: modules/renter/proto/merkleroots.go [hunk 23]
```diff
@@ -0,0 +1,359 @@
...
+	}
+	// If it is not uncached we need to delete the last cached tree and load
+	// its elements into mr.uncachedRoots.
+	mr.cachedSubTrees = mr.cachedSubTrees[:len(mr.cachedSubTrees)-1]
+	rootIndex := len(mr.cachedSubTrees) * merkleRootsPerCache
+	roots, err := mr.merkleRootsFromIndex(rootIndex, mr.numMerkleRoots)
+	if err != nil {
+		return errors.AddContext(err, "failed to read cached tree's roots")
+	}
+	mr.uncachedRoots = append(mr.uncachedRoots, roots...)
```
Comment (DavidVorick) [inline modules/renter/proto/merkleroots.go]: this logic of deleting the last cached and populating the uncached appears in a few places, probably worth combining them into a helper method.

File: modules/renter/proto/merkleroots.go (hunk 22 above)
Comment (ChrisSchinnerl) [reply modules/renter/proto/merkleroots.go]: Yeah that's what the `bufio.NewReader(file)` line should do. I'm not sure how much it actually buffers but it should do some good userspace level caching.  @lukechampine probably knows more about this.

File: modules/renter/proto/merkleroots.go [hunk 24]
```diff
@@ -0,0 +1,368 @@
...
+// len returns the number of merkle roots. It should always return the same
+// number as lenFromFile.
+func (mr *merkleRoots) len() int {
+	return mr.numMerkleRoots
+}
+
+// appendRoot appends a root to the in-memory structure of the merkleRoots. If
+// the length of the uncachedRoots grows too large they will be compressed into
+// a cachedSubTree.
+func (mr *merkleRoots) appendRoot(root crypto.Hash) {
```
Comment (DavidVorick) [inline modules/renter/proto/merkleroots.go]: may want to call this `appendRootMemory` or something, so that the function name clearly indicates that it only modifies the in-memory structure.

File: modules/renter/proto/merkleroots.go [hunk 25]
```diff
@@ -0,0 +1,368 @@
...
+	mr.appendRoot(root)
+
+	// Increment the number of roots.
+	mr.numMerkleRoots++
+	return nil
+}
+
+// root returns the root of the merkle roots.
+func (mr *merkleRoots) root() crypto.Hash {
+	tree := crypto.NewCachedTree(sectorHeight)
```
Comment (DavidVorick) [inline modules/renter/proto/merkleroots.go]: If we're working with pushing sub-roots, does it need to be a cached tree? Can it just be a normal tree?

I think that with the help of the `PushSubTree` function, we can stop using the `CachedTree` type.

File: modules/renter/proto/merkleroots.go [hunk 26]
```diff
@@ -0,0 +1,368 @@
...
+	}
+	for _, root := range mr.uncachedRoots {
+		tree.Push(root)
+	}
+	return tree.Root()
+}
+
+// newRoot returns the root of the merkleTree after appending the newRoot
+// without actually appending it.
+func (mr *merkleRoots) newRoot(newRoot crypto.Hash) crypto.Hash {
```
Comment (DavidVorick) [inline modules/renter/proto/merkleroots.go]: I would call this `checkNewRoot`

File: modules/renter/proto/contract.go [hunk 27]
```diff
@@ -96,14 +95,20 @@ type SafeContract struct {
...
 	// applied to the contract file.
 	unappliedTxns []*writeaheadlog.Transaction
 
+	// file is the file of the safe contract that contains the roots. This
+	// file is usually shared with the SafeContract which means multiple
+	// threads could potentially write to the same file. That's why the
+	// SafeContract should never modify the file beyond the
+	// contractHeaderSize and the merkleRoots should never modify data
+	// before that. Both should also use WriteAt and ReadAt instead of
+	// Write and Read.
```
Comment (lukechampine) [inline modules/renter/proto/contract.go]: can make this safer like so:
```go
type fileSection struct {
	f     *os.File
	start int64
	end   int64
}

func (f *fileSection) WriteAt(p []byte, off int64) (int, error) {
	return f.f.WriteAt(b, f.start+off)
}

// also implement ReadAt, Sync, and Close
```
The tricky part is handling out-of-bounds writes. For the header part of the file, you want to disallow writes past the end. But for the merkle roots part, you need to allow them. Two approaches come to mind:

1) Set `end: -1` for the merkle roots `fileSection`, and check for `f.end == -1` in `WriteAt`.
2) Explicitly extend `f.end` before calling `WriteAt`.

Option 2 seems a bit safer, but I won't make the final call.

The calls to `WriteAt` and `ReadAt` in the merkle roots section also need to updated to use the correct offsets.

File: modules/renter/proto/merkleroots.go [hunk 28]
```diff
@@ -0,0 +1,375 @@
...
+		tree.Push(root)
+	}
+	// Push the new root.
+	tree.Push(newRoot)
+	return tree.Root()
+}
+
+// merkleRoots reads all the merkle roots from disk and returns them. This is
+// not very fast and should only be used for testing purposes.
+func (mr *merkleRoots) merkleRoots() (roots []crypto.Hash, err error) {
```
Comment (DavidVorick) [inline modules/renter/proto/merkleroots.go]: If this function is only intended for testing, I would move it to an _test.go file, and then probably rename it to `merkleRootsFromDisk()`

File: modules/renter/proto/merkleroots.go [hunk 29]
```diff
@@ -0,0 +1,375 @@
...
+	}
+	// Sanity check: should have read exactly numMerkleRoots roots.
+	if len(roots) != mr.numMerkleRoots {
+		build.Critical(fmt.Sprintf("Number of merkle roots on disk (%v) doesn't match numMerkleRoots (%v)",
+			len(roots), mr.numMerkleRoots))
+	}
+	return
+}
+
+// merkleRootsFrom index readds all the merkle roots in range [from;to)
```
Comment (DavidVorick) [inline modules/renter/proto/merkleroots.go]: typo - readds

File: modules/renter/proto/merkleroots.go [hunk 30]
```diff
@@ -0,0 +1,375 @@
...
+func (mr *merkleRoots) merkleRootsFromIndex(from, to int) ([]crypto.Hash, error) {
```
Comment (DavidVorick) [inline modules/renter/proto/merkleroots.go]: would probably also have it be `merkleRootsFromIndexFromDisk`. Though that's a bit of a mouthful.

File: modules/renter/proto/contract.go [hunk 31]
```diff
@@ -411,17 +399,17 @@ func (cs *ContractSet) managedInsertContract(h contractHeader, roots []crypto.Ha
 	if err != nil {
 		return modules.RenterContract{}, err
 	}
-	// preallocate space for header + roots
-	if err := f.Truncate(contractHeaderSize + crypto.HashSize*int64(len(roots))); err != nil {
-		return modules.RenterContract{}, err
-	}
+	// create fileSections
+	headerSection := newFileSection(f, 0, contractHeaderSize)
+	rootsSection := newFileSection(f, contractHeaderSize+1, -1)
```
Comment (DavidVorick) [inline modules/renter/proto/contract.go]: my preference would be to change the `-1` sentinel to a constant. It's not super obvious, had to read the implementation to figure out what it did. But if it was a constant `remainingFile`, it might be more clear.

File: modules/renter/proto/filesection.go [hunk 32]
```diff
@@ -0,0 +1,98 @@
...
+	return f.f.ReadAt(b, f.start+off)
+}
+
+// Sync calls Sync on the fileSection's underlying file.
+func (f *fileSection) Sync() error {
+	return f.f.Sync()
+}
+
+// Truncate calls Truncate on the fileSection's underlying file.
+func (f *fileSection) Truncate(size int64) error {
```
Comment (DavidVorick) [inline modules/renter/proto/filesection.go]: should just return an error period if you call `Truncate` on a `fileSection` and `end != -1`

File: modules/renter/proto/merkleroots.go [hunk 33]
```diff
@@ -0,0 +1,369 @@
...
+func (mr *merkleRoots) appendRootMemory(root crypto.Hash) {
+	mr.uncachedRoots = append(mr.uncachedRoots, root)
+	if len(mr.uncachedRoots) == merkleRootsPerCache {
+		mr.cachedSubTrees = append(mr.cachedSubTrees, newCachedSubTree(mr.uncachedRoots))
+		mr.uncachedRoots = mr.uncachedRoots[:0]
+	}
+}
+
+// delete deletes the sector root at a certain index.
+func (mr *merkleRoots) delete(i int) error {
```
Comment (DavidVorick) [inline modules/renter/proto/merkleroots.go]: This operation isn't WAL-safe, it's not idempotent. If the WAL is going to be calling `delete` from a transaction, it needs to be able to call delete multiple times without causing harm.

the issue is that if we call `delete(5)` twice in a row, we delete two different roots, instead of deleting the same root. If this is to be called from the WAL, the instructions actually have to take 2 inputs:

1. some data to write at a specific index (for the swap) (might need a bool too indicating whether a swap is required)

2. some data indicating which index should be truncated to.

Those values combined will let you perform idempotent deletes.

File: modules/renter/proto/merkleroots.go [hunk 34]
```diff
@@ -0,0 +1,369 @@
...
+	if err != nil {
+		return errors.AddContext(err, "failed to read cached tree's roots")
+	}
+	mr.uncachedRoots = append(mr.uncachedRoots, roots...)
+	return nil
+}
+
+// push appends a merkle root to the end of the contract. If the number of
+// uncached merkle roots grows too big we cache them in a new subTree.
+func (mr *merkleRoots) push(root crypto.Hash) error {
```
Comment (DavidVorick) [inline modules/renter/proto/merkleroots.go]: `push` has the same problem as delete - it's not idempotent. WAL may end up calling an update multiple times if there are crashes at key places during disk management, so we need to restructure all of these functions to be idempotent.

This means we also need to write idempotency tests to verify that all of the operations are being updated idempotently. In the host we do this by adding a disrupt during WAL application that might write an update multiple times.

True idempotency though would be to apply some of the transactions in a WAL, perhaps even get partway through one of the transactions, and then apply them again a few times, each time stopping at a different update (or even finishing all the way), until finally on the last one you just go all the way.

If the WAL is built correctly, the result should be identical to just applying the transaction once.

File: modules/renter/proto/merkleroots.go [hunk 35]
```diff
@@ -66,27 +81,26 @@ func loadExistingMerkleRoots(file *fileSection) (*merkleRoots, error) {
...
-	}
 	// Read the roots from the file without reading all of them at once.
-	r := bytes.NewBuffer(fileData)
-	for i := 0; i < mr.numMerkleRoots; i++ {
-		var root crypto.Hash
-		if _, err = io.ReadFull(r, root[:]); err == io.EOF {
+	readOff := int64(0)
+	rootsData := make([]byte, rootsDiskLoadBulkSize)
+	for {
+		n, err := file.ReadAt(rootsData, readOff)
```
Comment (lukechampine) [inline modules/renter/proto/merkleroots.go]: are you not able to use `bufio` anymore because `file` only supports `ReadAt`, not `Read`?
That makes me sad.

File: modules/renter/proto/consts.go [hunk 36]
```diff
@@ -11,6 +11,14 @@ import (
 const (
 	// contractExtension is the extension given to contract files.
 	contractExtension = ".contract"
+
+	// rootsDiskLoadBulkSize is the max number of roots we read from disk at
+	// once to avoid using up all the ram.
+	rootsDiskLoadBulkSize = 1024 * crypto.HashSize // 32 kib
+
+	// remainingFile is a constant used to indicate that a fileSection can access
+	// the whole remaining file instead of being boung to a certain end offset.
```
Comment (DavidVorick) [inline modules/renter/proto/consts.go]: typo - boung

File: modules/renter/proto/contract.go [hunk 37]
```diff
@@ -447,27 +435,22 @@ func (cs *ContractSet) loadSafeContract(filename string, walTxns []*writeaheadlo
 	if err != nil {
 		return err
 	}
+	headerSection := newFileSection(f, 0, contractHeaderSize)
+	rootsSection := newFileSection(f, contractHeaderSize+1, -1)
```
Comment (DavidVorick) [inline modules/renter/proto/contract.go]: need to use the constant here

File: modules/renter/proto/filesection.go [hunk 38]
```diff
@@ -0,0 +1,94 @@
...
+	if f.end != remainingFile {
+		panic("can't truncate a file that has a end != remainingFile")
+	}
+	if f.start+size < f.start {
+		panic("can't truncate file to be smaller than the section start")
+	}
+	if f.start+size > f.end && f.end != remainingFile {
```
Comment (DavidVorick) [inline modules/renter/proto/filesection.go]: this block is unreachable, but also unneeded

File: modules/renter/proto/filesection.go [hunk 39]
```diff
@@ -0,0 +1,94 @@
+package proto
```
Comment (DavidVorick) [inline modules/renter/proto/filesection.go]: This file needs tests to prove correctness. Especially around all of the edge cases that it's supposed to protect against.
